agent/consul/state/acl_oss.go

```diff
@@ -0,0 +1,501 @@
+	return nil
+}
+
+func (s *Store) ACLPolicyUpsertValidateEnterprise(*structs.ACLPolicy, *structs.ACLPolicy) error {
+	return nil
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/////                        ACL Token Functions                          /////
+///////////////////////////////////////////////////////////////////////////////
+
+func (s *Store) aclTokenInsert(tx *memdb.Txn, token *structs.ACLToken, updateIndexes bool) error {
+	// insert the token into memdb
+	if err := tx.Insert("acl-tokens", token); err != nil {
+		return fmt.Errorf("failed inserting acl token: %v", err)
+	}
+
+	if updateIndexes {
+		// update the overall acl-tokens index
+		if err := tx.Insert("index", &IndexEntry{"acl-tokens", token.ModifyIndex}); err != nil {
+			return fmt.Errorf("failed updating acl tokens index: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (s *Store) aclTokenGetFromIndex(tx *memdb.Txn, id string, index string, entMeta *structs.EnterpriseMeta) (<-chan struct{}, interface{}, error) {
+	return tx.FirstWatch("acl-tokens", index, id)
+}
+
+func (s *Store) aclTokenListAll(tx *memdb.Txn, _ *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get("acl-tokens", "id")
+}
+
+func (s *Store) aclTokenListLocal(tx *memdb.Txn, _ *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get("acl-tokens", "local", true)
+}
+
+func (s *Store) aclTokenListGlobal(tx *memdb.Txn, _ *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get("acl-tokens", "local", false)
+}
+
+func (s *Store) aclTokenListByPolicy(tx *memdb.Txn, policy string, _ *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get("acl-tokens", "policies", policy)
+}
+
+func (s *Store) aclTokenListByRole(tx *memdb.Txn, role string, _ *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get("acl-tokens", "roles", role)
+}
+
+func (s *Store) aclTokenListByAuthMethod(tx *memdb.Txn, authMethod string, _ *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get("acl-tokens", "authmethod", authMethod)
+}
+
+func (s *Store) aclTokenDeleteWithToken(tx *memdb.Txn, token *structs.ACLToken, idx uint64) error {
+	// remove the token
+	if err := tx.Delete("acl-tokens", token); err != nil {
+		return fmt.Errorf("failed deleting acl token: %v", err)
+	}
+
+	// update the overall acl-tokens index
+	if err := tx.Insert("index", &IndexEntry{"acl-tokens", idx}); err != nil {
+		return fmt.Errorf("failed updating acl tokens index: %v", err)
+	}
+	return nil
+}
+
+func (s *Store) aclTokenMaxIndex(tx *memdb.Txn, _ *structs.ACLToken, entMeta *structs.EnterpriseMeta) uint64 {
+	return maxIndexTxn(tx, "acl-tokens")
+}
+
+func (s *Store) aclTokenUpsertValidateEnterprise(tx *memdb.Txn, token *structs.ACLToken, existing *structs.ACLToken) error {
+	return nil
+}
+
+func (s *Store) ACLTokenUpsertValidateEnterprise(token *structs.ACLToken, existing *structs.ACLToken) error {
```

Comment:
Why do both `aclTokenUpsertValidateEnterprise` and `ACLTokenUpsertValidateEnterprise` exist?

Reply:
In enterprise these will validate that the namespace of the token (or other acl resource is valid and unchanging). The unexported one is used within the state package and the exported ones are used by the RPC functions.